csrc/device_lower/pass/circular_buffer.cpp

```diff
@@ -215,6 +216,706 @@ class CircularBufferLoopCloner : public kir::IrVisitor {
+  //    }
+  //  } else {
+  //    tokens[next_stage] = mbarrier::arrive(mbarrier[next_stage]);
+  //  }
+  void addTmaLoadBlock(Expr* expr) {
+    NVF_ERROR(mbarrier_arrive_tx_ != nullptr);
+    NVF_ERROR(expr != nullptr);
+
+    // Create the if-then-else with electSync() predicat for the arrive expect
+    // transaction.
+    kir::IfThenElse* if_expr = IrBuilder::create<kir::IfThenElse>(
+        IrBuilder::create<kir::Predicate>(PredicateType::ElectSync));
+
+    // A single thread issues arriveExpectTx with expected transactions and
+    // launches the TMA load.
+    if_expr->thenBody().push_back(mbarrier_arrive_tx_);
+    if_expr->thenBody().push_back(expr);
+
+    // The other threads issue arriveExpectTx without any expected transactions.
+    kir::MBarrierArrive* thread_arrive = IrBuilder::create<kir::MBarrierArrive>(
+        mbarrier_arrive_tx_->state(), mbarrier_arrive_tx_->mbarrier());
+    if_expr->elseBody().push_back(thread_arrive);
+    for_loop_stack_.back()->body().push_back(if_expr);
+
+    mbarrier_arrive_tx_ = nullptr;
+  }
+
+  // Get size of tma load in bytes. It is used for expected transaction count in
+  // kir::MBarrierArriveExpectTx.
+  Val* getSizeOfTmaLoad(LoadStoreOp* ldst) {
+    NVF_ERROR(ldst != nullptr);
+
+    TensorView* consumer_tv = ldst->out()->as<TensorView>();
+    NVF_ERROR(
+        GpuLower::current()->consumerToTMAInfo().count(consumer_tv),
+        "Unable to find TMA info for consumer_tv: ",
+        consumer_tv->toString());
+
+    // Get expected bytes for given TMA load operation.
+    const TMAInfo& tma_info =
+        GpuLower::current()->consumerToTMAInfo().at(consumer_tv);
+    Val* expected_bytes = tma_info.tileSizeBytes();
+
+    // The expected_bytes for mbarrier::arriveExpectTX must account for all TMA
+    // load operations launched for each circular buffer stage. We take the
+    // product of all coordinate TMA iterDomains to the right of the circular
+    // buffer axis.
+    const std::vector<IterDomain*>& leaf_domain = consumer_tv->getLoopDomain();
+    for (size_t idx = consumer_tv->getComputeAtPosition();
+         idx < leaf_domain.size();
+         ++idx) {
+      IterDomain* id = leaf_domain.at(idx);
+      if (!isParallelTypeThread(id->getParallelType()) &&
+          id->getParallelType() != ParallelType::Bulk) {
+        expected_bytes =
+            SimplifyingIrBuilder::mulExpr(expected_bytes, id->extent());
+      }
+    }
+    expected_bytes =
+        SimplifyingIrBuilder::maybeCastExpr(DataType::UInt32, expected_bytes);
+
+    auto is_multiple_of_16B = SimplifyingIrBuilder::eqExpr(
+        SimplifyingIrBuilder::modExpr(
+            expected_bytes, IrBuilder::create<Val>(16, DataType::Index)),
+        expected_bytes->fusion()->zeroVal());
+    GpuLower::current()->validate(
+        is_multiple_of_16B,
+        "The expected bytes must be a multiple of 16 bytes, but ",
+        expected_bytes,
+        " is not.");
+    return expected_bytes;
```

Comment:
Is this required? We are already validating that a single TMA instruction transfers a multiple of 16B in the code below:
 
https://github.com/NVIDIA/Fuser/blob/7b2c77c3d700eb918c8c7944d37bcf454c82f2ac/csrc/index_compute.cpp#L2683-L2695

I think this is the only requirement from hardware.

And if `tileSizeBytes()` is a multiple of 16 bytes, then the `expected_bytes` is guaranteed to be a multiple of 16 bytes mathematically.

Reply:
I removed it.